File: src/engine/model/Condition.cpp
/*
 * Condition.cpp
 *
 *  Created on: Mar 5, 2014
 *      Author: Stephan Opfer
 */

#include "engine/model/Condition.h"
#include "engine/model/AbstractPlan.h"
#include "engine/model/Quantifier.h"

namespace alica
{

	Condition::Condition()
	{
	}

	Condition::Condition(long id)
	{
		this->id = id;
		this->abstractPlan = NULL;
	}

	Condition::~Condition()
	{
	}

	const string& Condition::getConditionString() const
	{
		return conditionString;
	}

	void Condition::setConditionString(const string& conditionString)
	{
		this->conditionString = conditionString;
	}

	list<Quantifier*>& Condition::getQuantifiers()
	{
		return quantifiers;
	}

	void Condition::setQuantifiers(const list<Quantifier*>& quantifiers)
	{
		this->quantifiers = quantifiers;
	}

	const list<Variable*>& Condition::getVariables() const
	{
		return variables;
	}

	void Condition::setVariables(const list<Variable*>& variables)
	{
		this->variables = variables;
	}

	AbstractPlan* Condition::getAbstractPlan() const
	{
		return abstractPlan;
	}

	void Condition::setAbstractPlan(AbstractPlan* abstractPlan)
	{
		this->abstractPlan = abstractPlan;
	}

} /* namespace Alica */

File: supplementary/alica_codegeneration/alica-plan-designer-fx-model/src/main/java/de/unikassel/vs/alica/planDesigner/alicamodel/Condition.java
package de.unikassel.vs.alica.planDesigner.alicamodel;

import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Condition extends PlanElement {

    protected final SimpleBooleanProperty enabled = new SimpleBooleanProperty(this, "enabled", false);
    protected final SimpleStringProperty conditionString = new SimpleStringProperty(this, "conditionString", null);
    protected final SimpleStringProperty pluginName = new SimpleStringProperty(this, "pluginName", null);
    protected final ArrayList<Variable> variables = new ArrayList<>();
    protected final ArrayList<Quantifier> quantifiers = new ArrayList<>();
    private ChangeListenerForDirtyFlag listenerForDirtyFlag;

    public boolean getEnabled () {
        return enabled.get();
    }
    public void setEnabled(boolean enabled) {
        this.enabled.set(enabled);
    }
    public SimpleBooleanProperty enabledProperty() {
        return enabled;
    }

    public String getConditionString() {
        return conditionString.get();
    }
    public void setConditionString(String conditionString) {
        this.conditionString.set(conditionString);
    }
    public SimpleStringProperty conditionStringProperty() {
        return conditionString;
    }

    public String getPluginName() {
        return pluginName.get();
    }
    public void setPluginName(String pluginName) {
        this.pluginName.set(pluginName);
    }
    public SimpleStringProperty pluginNameProperty() {
        return pluginName;
    }

    public List<Variable> getVariables() {
        return Collections.unmodifiableList(variables);
    }
    public void addVariable(Variable variable){
        variables.add(variable);
        if (listenerForDirtyFlag != null) {
            listenerForDirtyFlag.setDirty();
        }
    }
    public void removeVariable(Variable variable){
        variables.remove(variable);
        if (listenerForDirtyFlag != null) {
            listenerForDirtyFlag.setDirty();
        }
    }

    public List<Quantifier> getQuantifiers() {
        return Collections.unmodifiableList(quantifiers);
    }
    public void addQuantifier(Quantifier quantifier) {
        quantifiers.add(quantifier);
        quantifier.registerDirtyFlag(listenerForDirtyFlag);
        if (listenerForDirtyFlag != null) {
            listenerForDirtyFlag.setDirty();
        }
    }
    public void removeQuantifier(Quantifier quantifier) {
        quantifiers.remove(quantifier);
        if (listenerForDirtyFlag != null) {
            listenerForDirtyFlag.setDirty();
        }
    }

    public void registerDirtyFlag(ChangeListenerForDirtyFlag listener) {
        if (listener == null) {
            return;
        }
        this.listenerForDirtyFlag = listener;

        this.name.addListener(listener);
        this.comment.addListener(listener);
        this.enabled.addListener(listener);
        this.conditionString.addListener(listener);
        this.pluginName.addListener(listener);

        for (Quantifier quantifier : quantifiers) {
            quantifier.registerDirtyFlag(listener);
        }
        // Variables are owned by an AbstractPlan, so listeners are registered there and exist already
    }

    public boolean hasConstraint() {
        return !variables.isEmpty() || !quantifiers.isEmpty();
    }
}